server/sonar-server/src/main/java/org/sonar/server/authentication/OAuth2CallbackFilter.java

```diff
@@ -61,34 +68,75 @@ public UrlPattern doGetPattern() {
   @Override
   public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
     HttpServletRequest httpRequest = (HttpServletRequest) request;
-    String requestUri = httpRequest.getRequestURI();
-    String keyProvider = "";
+    HttpServletResponse httpResponse = (HttpServletResponse) response;
+
+    IdentityProvider provider = resolveProviderOrHandleResponse(httpRequest, httpResponse);
+    if (provider != null) {
+      handleProvider(httpRequest, (HttpServletResponse) response, provider);
+    }
+  }
+
+  @CheckForNull
+  private IdentityProvider resolveProviderOrHandleResponse(HttpServletRequest request, HttpServletResponse response) {
+    String requestUri = request.getRequestURI();
+    String providerKey = extractKeyProvider(requestUri, server.getContextPath() + CALLBACK_PATH);
+    if (providerKey == null) {
+      handleError(response, "No provider key found in URI");
+      return null;
+    }
     try {
-      keyProvider = extractKeyProvider(requestUri, server.getContextPath() + CALLBACK_PATH);
-      IdentityProvider provider = identityProviderRepository.getEnabledByKey(keyProvider);
-      if (provider instanceof OAuth2IdentityProvider) {
-        OAuth2IdentityProvider oauthProvider = (OAuth2IdentityProvider) provider;
-        oauthProvider.callback(oAuth2ContextFactory.newCallback(httpRequest, (HttpServletResponse) response, oauthProvider));
-      } else {
-        handleError((HttpServletResponse) response, format("Not an OAuth2IdentityProvider: %s", provider.getClass()));
-      }
-    } catch (UnauthorizedException e) {
-      handleUnauthorizedError(e, (HttpServletResponse) response);
-    } catch (Exception e) {
-      handleError(e, (HttpServletResponse) response,
-        keyProvider.isEmpty() ? "Fail to callback authentication" :
-          format("Fail to callback authentication with '%s'", keyProvider));
+      return identityProviderRepository.getEnabledByKey(providerKey);
+    } catch (IllegalArgumentException e) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1166)